src/order/closure.lean

```diff
@@ -158,70 +170,264 @@ lemma eq_mk₃_closed (c : closure_operator α) :
-operator.
+lemma idempotent (x : α) : u (l (u (l x))) = u (l x) :=
+l.closure_operator.idempotent _
 
-Note that the inverse in the opposite direction does not hold in general.
--/
+lemma le_closure_iff (x y : α) : x ≤ u (l y) ↔ u (l x) ≤ u (l y) :=
+l.closure_operator.le_closure_iff _ _
+
+end partial_order
+
+section preorder
+variables [preorder α] [preorder β] {u : β → α} (l : lower_adjoint u)
+
+/-- An element `x` is closed for the closure operator `c` if it is a fixed point for it. -/
+def closed : set α := λ x, u (l x) = x
+
+lemma mem_closed_iff (x : α) : x ∈ l.closed ↔ u (l x) = x := iff.rfl
+
+lemma closure_eq_self_of_mem_closed {x : α} (h : x ∈ l.closed) : u (l x) = x := h
+
+end preorder
+
+section partial_order
+variables [partial_order α] [partial_order β] {u : β → α} (l : lower_adjoint u)
+
+lemma mem_closed_iff_closure_le (x : α) : x ∈ l.closed ↔ u (l x) ≤ x :=
+l.closure_operator.mem_closed_iff_closure_le _
+
+@[simp] lemma closure_is_closed (x : α) : u (l x) ∈ l.closed := l.idempotent x
+
+/-- The set of closed elements for `c` is exactly its range. -/
+lemma closed_eq_range_close : l.closed = set.range (u ∘ l) :=
+l.closure_operator.closed_eq_range_close
+
+/-- Send an `x` to an element of the set of closed elements (by taking the closure). -/
+def to_closed (x : α) : l.closed := ⟨u (l x), l.closure_is_closed x⟩
+
+@[simp] lemma closure_le_closed_iff_le (x : α) {y : α} (hy : l.closed y) : u (l x) ≤ y ↔ x ≤ y :=
+l.closure_operator.closure_le_closed_iff_le x hy
+
+end partial_order
+
+lemma closure_top [order_top α] [preorder β] {u : β → α} (l : lower_adjoint u) :
+  u (l ⊤) = ⊤ :=
+l.closure_operator.closure_top
+
+lemma closure_inf_le [semilattice_inf α] [preorder β] {u : β → α} (l : lower_adjoint u) (x y : α) :
+  u (l (x ⊓ y)) ≤ u (l x) ⊓ u (l y) :=
+l.closure_operator.closure_inf_le x y
+
+section semilattice_sup
+variables [semilattice_sup α] [preorder β] {u : β → α} (l : lower_adjoint u)
+
+lemma closure_sup_closure_le (x y : α) :
+  u (l x) ⊔ u (l y) ≤ u (l (x ⊔ y)) :=
+l.closure_operator.closure_sup_closure_le x y
+
+@[simp] lemma closure_sup_closure_left (x y : α) :
+  u (l (u (l x) ⊔ y)) = u (l (x ⊔ y)) :=
+l.closure_operator.closure_sup_closure_left x y
+
+@[simp] lemma closure_sup_closure_right (x y : α) :
+  u (l (x ⊔ u (l y))) = u (l (x ⊔ y)) :=
+l.closure_operator.closure_sup_closure_right x y
+
+@[simp] lemma closure_sup_closure (x y : α) :
+  u (l (u (l x) ⊔ u (l y))) = u (l (x ⊔ y)) :=
+l.closure_operator.closure_sup_closure x y
+
+end semilattice_sup
+
+section complete_lattice
+variables [complete_lattice α] [preorder β] {u : β → α} (l : lower_adjoint u)
+
+@[simp] lemma closure_supr_closure {ι : Type u} (x : ι → α) :
+  u (l (⨆ i, u (l (x i)))) = u (l (⨆ i, x i)) :=
+l.closure_operator.closure_supr_closure x
+
+@[simp] lemma closure_bsupr_closure (p : α → Prop) :
+  u (l (⨆ x (H : p x), u (l x))) = u (l (⨆ x (H : p x), x)) :=
+l.closure_operator.closure_bsupr_closure p
```

Comment:
These lemmas cannot be marked `simp` because the head symbol (the left-most function application on the left hand side) is a variable `u`, and `simp` needs this to be the name of a declaraton. I expect the best thing to do here is to duplicate them in the `coe_to_set` section, setting `u := coe`. (See also the results of the `simp` linter.)

```suggestion
lemma closure_sup_closure_left (x y : α) :
  u (l (u (l x) ⊔ y)) = u (l (x ⊔ y)) :=
l.closure_operator.closure_sup_closure_left x y

lemma closure_sup_closure_right (x y : α) :
  u (l (x ⊔ u (l y))) = u (l (x ⊔ y)) :=
l.closure_operator.closure_sup_closure_right x y

lemma closure_sup_closure (x y : α) :
  u (l (u (l x) ⊔ u (l y))) = u (l (x ⊔ y)) :=
l.closure_operator.closure_sup_closure x y

end semilattice_sup

section complete_lattice
variables [complete_lattice α] [preorder β] {u : β → α} (l : lower_adjoint u)

lemma closure_supr_closure {ι : Type u} (x : ι → α) :
  u (l (⨆ i, u (l (x i)))) = u (l (⨆ i, x i)) :=
l.closure_operator.closure_supr_closure x

lemma closure_bsupr_closure (p : α → Prop) :
  u (l (⨆ x (H : p x), u (l x))) = u (l (⨆ x (H : p x), x)) :=
l.closure_operator.closure_bsupr_closure p
```

Reply:
How should the doubles be named, then?

Reply:
Since they are all `set`-related, I would go for the same but with `set`-operators rather than `lattice`: `closure_union_closure_left`, `closure_Union_closure`, etc.

By the way, since `coe` is injective, we can get rid of the outermost `u`/`coe` in the doubles.